File: boj/FloydWarshall/BOJ_1058.py
# 백준 1058 친구
# 21.03.03

import sys
import heapq
from functools import cmp_to_key
from collections import deque
from itertools import combinations
from itertools import permutations
sys.setrecursionlimit(10**6)

# 거리가 2이하인 친구만 세기
INF = int(1e9)
n = int(input())
graph = [[INF] * n for _ in range(n)]

for i in range(n):
    line = sys.stdin.readline().rstrip()
    for j in range(n):
        if line[j] == 'Y':
            graph[i][j] = 1

for k in range(n):
    for i in range(n):
        for j in range(n):
            if i == j or i == k or j == k: continue
            graph[i][j] = min(graph[i][j], graph[i][k] + graph[k][j])

ans = 0
for i in range(n):
    cnt = 0
    for j in range(n):
        if graph[i][j] <= 2: cnt += 1
    ans = max(ans, cnt)

print(ans)

File: boj/FloydWarshall/BOJ_1058.swift
// 백준 1058 친구
// 21.03.27

import Foundation

let n = Int(readLine()!)!

var graph = [[Character]](repeating: [Character](), count: n)

for i in 0..<n {
  let arr = Array(readLine()!)
  
  graph[i] = arr
}

var ans = 0

for i in 0..<n {
  var cnt = 0
  
  for j in 0..<n {
    if i == j { continue }
    
    if graph[i][j] == "Y" {
      cnt += 1
      continue
    }
    
    // i, j가 친구가 아니면 2-친구 관계인지 확인
    for k in 0..<n {
      if graph[i][k] == "Y" && graph[k][j] == "Y" {
        cnt += 1
        break // 2-친구 관계인거 확인했으면 더이상 찾지 않아도 됨
      }
    }
  }
  ans = max(ans, cnt)
}

print(ans)
